lib/async_hooks.js

```diff
@@ -209,11 +211,102 @@ class AsyncResource {
   }
 }
 
+const storageList = [];
+const storageHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const currentResource = executionAsyncResource();
+    // Value of currentResource is always a non null object
+    for (let i = 0; i < storageList.length; ++i) {
+      storageList[i]._propagate(resource, currentResource);
+    }
+  }
+});
+
+class AsyncLocalStorage {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+    this.enabled = false;
+  }
+
+  disable() {
+    if (this.enabled) {
+      this.enabled = false;
+      // If this.enabled, the instance must be in storageList
+      storageList.splice(storageList.indexOf(this), 1);
+      if (storageList.length === 0) {
+        storageHook.disable();
+      }
+    }
+  }
+
+  // Propagate the context from a parent resource to a child one
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (this.enabled) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  _enter() {
+    if (!this.enabled) {
+      this.enabled = true;
+      storageList.push(this);
+      storageHook.enable();
+    }
+    const resource = executionAsyncResource();
+    resource[this.kResourceStore] = new Map();
+  }
+
+  _exit() {
+    const resource = executionAsyncResource();
+    if (resource) {
+      resource[this.kResourceStore] = undefined;
+    }
+  }
+
+  runSyncAndReturn(callback, ...args) {
+    this._enter();
+    try {
+      return callback(...args);
+    } finally {
+      this._exit();
+    }
+  }
+
+  exitSyncAndReturn(callback, ...args) {
+    this.enabled = false;
+    try {
+      return callback(...args);
+    } finally {
+      this.enabled = true;
+    }
+  }
+
+  getStore() {
+    const resource = executionAsyncResource();
+    if (this.enabled) {
+      return resource[this.kResourceStore];
+    }
+  }
+
+  run(callback, ...args) {
+    this._enter();
+    process.nextTick(() => callback(...args));
```

Comment:
Very minor nit, but this feels cleaner/simpler to me.

```suggestion
    process.nextTick(callback, ...args);
```

Reply:
Addressed in https://github.com/vdeturckheim/node/pull/4